src/Tutor.Core/UseCases/Learning/SessionService.cs

```diff
@@ -50,4 +50,55 @@ public Result TerminateSession(int knowledgeComponentId, int learnerId)
 
         return result;
     }
+
+    public Result PauseSession(int knowledgeComponentId, int learnerId)
+    {
+        if (!_enrollmentRepository.HasActiveEnrollmentForKc(knowledgeComponentId, learnerId))
+            return Result.Fail(FailureCode.NotEnrolledInUnit);
+
+        var kcMastery = _knowledgeMasteryRepository.GetBareKcMastery(knowledgeComponentId, learnerId);
+        if (kcMastery == null) return Result.Fail(FailureCode.NotFound);
+
+        var result = kcMastery.PauseSession();
+        if (result.IsFailed) return result;
+        _knowledgeMasteryRepository.UpdateKcMastery(kcMastery);
+        result = _unitOfWork.Save();
+        if (result.IsFailed) return result;
+
+        return result;
+    }
+
+    public Result ContinueSession(int knowledgeComponentId, int learnerId)
+    {
+        if (!_enrollmentRepository.HasActiveEnrollmentForKc(knowledgeComponentId, learnerId))
+            return Result.Fail(FailureCode.NotEnrolledInUnit);
+
+        var kcMastery = _knowledgeMasteryRepository.GetBareKcMastery(knowledgeComponentId, learnerId);
+        if (kcMastery == null) return Result.Fail(FailureCode.NotFound);
+
+        var result = kcMastery.ContinueSession();
+        if (result.IsFailed) return result;
+        _knowledgeMasteryRepository.UpdateKcMastery(kcMastery);
+        result = _unitOfWork.Save();
+        if (result.IsFailed) return result;
+
+        return result;
+    }
+
+    public Result AbandonSession(int knowledgeComponentId, int learnerId)
+    {
+        if (!_enrollmentRepository.HasActiveEnrollmentForKc(knowledgeComponentId, learnerId))
+            return Result.Fail(FailureCode.NotEnrolledInUnit);
+
+        var kcMastery = _knowledgeMasteryRepository.GetBareKcMastery(knowledgeComponentId, learnerId);
+        if (kcMastery == null) return Result.Fail(FailureCode.NotFound);
+
+        var result = kcMastery.AbandonSession();
+        if (result.IsFailed) return result;
+        _knowledgeMasteryRepository.UpdateKcMastery(kcMastery);
+        result = _unitOfWork.Save();
+        if (result.IsFailed) return result;
+
+        return result;
+    }
```

Comment:
Can we simplify this whole service?
Each method does the same checks, just generates different event.
Should we extract code into two private methods: before and after boiler plate code; and just keep the main logic in the methods?

Reply:
Yes, good catch.